cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -301,7 +301,7 @@ protected void postProcess() {
         }
         recordedRunRecord =
           appMetadataStore.recordProgramStop(programRunId, endTimeSecs, programRunStatus, null, messageIdBytes);
-
+        writeToHeartBeatDataset(recordedRunRecord, endTimeSecs, datasetContext, programHeartbeatDataset);
```

Comment:
Seems like in every switch case you are writing to heart beat dataset. Possible to do it outside of the switch?

Reply:
the timestamp changes based on the status, endtTme for end status, logicalStartTime in running,etc. since the timestamp value is available inside each case. i call it from inside.

Reply:
other option is using runRecord and its program status and getting timestamp based on that program status, if thats better.

Reply:
hum.. never mind for now. That big switch block deserves some refactoring later.